### PR TITLE
blockdev_snapshot: add nbd support for live snapshot

### PR DESCRIPTION
**Pull request by aliang123**
Add backend:nbd support for live snapshot

Signed-off-by: Aihua Liang <aliang@redhat.com>
id:2066584

### CODE REVIEW

**Comment by aliang123** [comment]
(01/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_stress.q35: PASS (80.43 s)
(02/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_merge.q35: PASS (97.12 s)
(03/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_data_file.q35: PASS (79.95 s)
(04/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_reboot.q35: PASS (77.07 s)
(05/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_readonly.q35: PASS (75.89 s)
(06/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot.q35: PASS (83.55 s)
(07/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_512.q35: PASS (83.14 s)
(08/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_2M.q35: PASS (81.28 s)
(09/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_guest_agent.q35: PASS (129.42 s)
(10/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_chains.q35: PASS (160.58 s)
(11/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_stop_cont.q35: PASS (57.09 s)
(12/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_multi_disks.q35: PASS (101.56 s)
(13/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_install.q35: PASS (630.07 s)
(14/14) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_nonexist_target.q35: PASS (79.25 s)

**Comment by aliang123** [comment]
@zhencliu @ngu-niny or @MiriamDeng, please help to review, thanks.

File: qemu/tests/cfg/blockdev_snapshot_install.cfg [hunk 1]
```diff
@@ -51,3 +51,11 @@
        enable_ceph_unattended = no
        enable_ceph_sn1 = no
        enable_ceph_image1 = yes
+    nbd:
+       enable_nbd_cd1 = no
+       enable_nbd_unattended = no
+       enable_nbd_sn1 = no
+       enable_nbd_image1 = yes
+       nbd_port_image1 = 10811
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_snapshot_install.cfg:59]: Do we want to add image_format_image1 = raw here to limit the target installation image format?

File: qemu/tests/cfg/blockdev_snapshot.cfg [hunk 2]
```diff
@@ -41,3 +41,12 @@
         enable_ceph_sn1 = no
         enable_ceph_data = yes
         image_format_data = raw
+    nbd:
+        enable_nbd_sn1 = no
+        enable_nbd_data = yes
+        image_format_data = raw
+        nbd_port_data = 10810
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_snapshot.cfg:48]: Discussed in the nbd setup doc, we may also need to update the ports used in all test cases

File: qemu/tests/blockdev_snapshot_chains.py [hunk 3]
```diff
@@ -26,17 +26,15 @@ def prepare_snapshot_file(self):
...
             self.params["image_format_%s" % snapshot_tag] = snapshot_format
-            if self.params["image_backend"] == "iscsi_direct":
+            backend = self.params["image_backend"]
+            if backend == "iscsi_direct":
                 self.params.update({"enable_iscsi_%s" % snapshot_tag: "no"})
                 self.params.update({"image_raw_device_%s" % snapshot_tag: "no"})
-            elif self.params["image_backend"] == "ceph":
-                self.params.update({"enable_ceph_%s" % snapshot_tag: "no"})
+            else:
+                self.params.update({"enable_%s_%s" % (backend, snapshot_tag): "no"})
```
Comment (zhencliu) [inline qemu/tests/blockdev_snapshot_chains.py:37]: Your update can use short code to implement function with the ceph/nbd backend, but it's not that clear to understand the codes, and if we wants to support more backends or if we want to do more settings for a specific backend, we may also have to update the codes here. So what about coding here in the same way as we did for iscsi_direct?

File: qemu/tests/blockdev_snapshot_merge.py [hunk 4]
```diff
@@ -21,11 +21,12 @@ def prepare_snapshot_file(self):
...
         self.params["image_format_%s" % self.snapshot_tag] = snapshot_format
-        if self.params["image_backend"] == "iscsi_direct":
+        backend = self.params["image_backend"]
+        if backend == "iscsi_direct":
             self.params.update({"enable_iscsi_%s" % self.snapshot_tag: "no"})
             self.params.update({"image_raw_device_%s" % self.snapshot_tag: "no"})
-        elif self.params["image_backend"] == "ceph":
-            self.params.update({"enable_ceph_%s" % self.snapshot_tag: "no"})
+        else:
+            self.params.update({"enable_%s_%s" % (backend, self.snapshot_tag): "no"})
```
Comment (zhencliu) [inline qemu/tests/blockdev_snapshot_merge.py:29]: same as above

**Comment by aliang123** [comment]
(1/3) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_merge.q35: PASS (98.68 s)
(2/3) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_chains.q35: PASS (165.08 s)
(3/3) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_multi_disks.q35: PASS (89.09 s)

File: qemu/tests/cfg/blockdev_snapshot_install.cfg (hunk 1 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_snapshot_install.cfg]: No need to add it, we set image_format=raw in our kar cmdline for system disk.

File: qemu/tests/cfg/blockdev_snapshot.cfg (hunk 2 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_snapshot.cfg]: Have updated all of them, thanks zhenchao.

File: qemu/tests/blockdev_snapshot_chains.py (hunk 3 above)
Comment (aliang123) [reply qemu/tests/blockdev_snapshot_chains.py]: Have upated it, thanks zhenchao.

File: qemu/tests/blockdev_snapshot_merge.py (hunk 4 above)
Comment (aliang123) [reply qemu/tests/blockdev_snapshot_merge.py]: Have updated it, thanks zhenchao.

File: qemu/tests/cfg/blockdev_snapshot_multi_disks.cfg [hunk 5]
```diff
@@ -44,3 +48,20 @@
...
+        enable_nbd_data1 = yes
+        enable_nbd_data2 = yes
+        image_format_data1 = raw
+        image_format_data2 = raw
+        nbd_port_data1 = 10831
+        nbd_port_data2 = 10832
+        image_size_data1 = 2G
+        image_size_data2 = 2G
+        image_size_sn1 = ${image_size_data1}
+        image_size_sn2 = ${image_size_data2}
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_snapshot_multi_disks.cfg:63]: Can we update line8&15 with the image size, then we can remove these 4 lines

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m9.u1.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_snapshot_chains.q35: PASS (247.12 s)

File: qemu/tests/cfg/blockdev_snapshot_multi_disks.cfg (hunk 5 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_snapshot_multi_disks.cfg]: After discussion with zhenchao, we decided to keep the original design.

**Comment by MiriamDeng** [comment]
Test passed on ppc,
 (01/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_merge: STARTED
\ (01/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_merge: PASS (161.36 s)
 (02/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_data_file: STARTED
| (02/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_data_file: PASS (108.44 s)
 (03/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_install: STARTED
| (03/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_install: FAIL: Failed to install guest (123.70 s)  
 (04/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_stop_cont: STARTED
\ (04/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_stop_cont: PASS (94.14 s)
 (05/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_readonly: STARTED
/ (05/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_readonly: PASS (117.52 s)
 (06/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_nonexist_target: STARTED
- (06/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_nonexist_target: PASS (51.53 s)
 (07/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot: STARTED
/ (07/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot: FAIL: AssertionError("Get file ('data') MD5 with error: md5sum: /mnt/sda1/data: No such file or directory\n",) (110.12 s)
 (08/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_512: STARTED
- (08/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_512: PASS (124.43 s)
 (09/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_2M: STARTED
| (09/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot.dst_cluster_size_2M: PASS (112.93 s)
 (10/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_stress: STARTED
/ (10/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_stress: PASS (108.05 s)
 (11/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_multi_disks: STARTED
/ (11/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_multi_disks: PASS (135.41 s)
 (12/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_chains: STARTED
- (12/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_chains: PASS (197.04 s)
 (13/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_reboot: STARTED
\ (13/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_reboot: PASS (100.62 s)
 (14/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_guest_agent: STARTED
\ (14/14) Host_RHEL.m8.u7.nbd.raw.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_snapshot_guest_agent: PASS (256.15 s)

**Comment by zhencliu** [comment]
LGTM

**Pull request merged**